File: app/src/main/java/com/sofps/inspirationalquotes/data/source/remote/QuotesRemoteDataSource.java
package com.sofps.inspirationalquotes.data.source.remote;

import androidx.annotation.NonNull;

import com.sofps.inspirationalquotes.data.Quote;
import com.sofps.inspirationalquotes.data.QuoteApi;
import com.sofps.inspirationalquotes.data.QuotesService;
import com.sofps.inspirationalquotes.data.source.QuotesDataSource;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class QuotesRemoteDataSource implements QuotesDataSource {

    private final static String LANGUAGE_SUPPORTED = "EN";

    private final QuotesService mQuotesService;

    public QuotesRemoteDataSource(QuotesService quotesService) {
        mQuotesService = quotesService;
    }

    @Override
    public void getQuotes(@NonNull String language, @NonNull final GetQuoteCallback callback) {
        if (LANGUAGE_SUPPORTED.equalsIgnoreCase(language)) {
            mQuotesService.getQuoteOfTheDay().enqueue(new Callback<QuoteApi>() {
                @Override
                public void onResponse(Call<QuoteApi> call, Response<QuoteApi> response) {
                    QuoteApi quoteApi = response.body();
                    if (response.isSuccessful() && quoteApi != null && quoteApi.isValid()) {
                        Quote quote = mapQuote(quoteApi);
                        List<Quote> quotes = new ArrayList<>(1);
                        quotes.add(quote);
                        callback.onQuotesLoaded(quotes);
                    } else {
                        callback.onDataNotAvailable();
                    }
                }

                @Override
                public void onFailure(Call<QuoteApi> call, Throwable t) {
                    callback.onDataNotAvailable();
                }
            });
        } else {
            callback.onDataNotAvailable();
        }
    }

    private Quote mapQuote(@NonNull QuoteApi quoteApi) {
        Quote quote = new Quote();
        quote.setLanguage(LANGUAGE_SUPPORTED);
        quote.setAuthor(quoteApi.getAuthor());
        quote.setText(quoteApi.getQuote());
        return quote;
    }
}

File: app/src/main/java/com/sofps/inspirationalquotes/data/source/remote/QuotesRemoteDataSource.kt
package com.sofps.inspirationalquotes.data.source.remote

import com.sofps.inspirationalquotes.data.QuoteApi
import com.sofps.inspirationalquotes.data.QuotesService
import com.sofps.inspirationalquotes.data.source.QuotesDataSource
import com.sofps.inspirationalquotes.data.QuoteDb
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import timber.log.Timber

class QuotesRemoteDataSource(private val quotesService: QuotesService) : QuotesDataSource {

    companion object {

        private const val LANGUAGE_SUPPORTED = "EN"
    }

    override fun getQuotes(language: String): Flow<List<QuoteDb>> {
        return flow {
            val quoteApi = quotesService.getQuoteOfTheDay()
            Timber.d("quoteApi=$quoteApi")
            emit(listOf(mapQuote(quoteApi)))
        }.flowOn(Dispatchers.IO)
    }

    private fun mapQuote(quoteApi: QuoteApi) =
            QuoteDb(language = LANGUAGE_SUPPORTED, author = quoteApi.author, text = quoteApi.quote)

}
